library/psa_crypto.c

```diff
@@ -3244,6 +3247,9 @@ psa_status_t psa_sign_hash_complete(
                                                     signature_length );
 exit:
 
+    /* Update ops count with work done. */
+    operation->num_ops += psa_driver_wrapper_sign_hash_get_num_ops( operation );
```

Comment:
As was I. 

The `ops_done` member of `mbedtls_ecp_restart_ctx` is reset to zero at the lowest level of any restartable call in the `ECP_RS_ENTER()` and `ECDSA_RS_ENTER()` macros. My initial implementation had this just to be set to equal the restart context value, but then the tests to see if the value kept going up after work kept failing, and lead me to discover this.

Reply:
Ah, yes, I had completely forgotten but now that you mention it, it makes sense: the `ops_done` member of `ecp_restart_ctx` is just there to count how many ops we've done this time round in order to decide whether we can take the next chunk of work of need to return now.

So, indeed the `+=` is quite correct, but perhaps a comment would be helpful, considering we were both surprised.

Reply:
I think beyond the absence of comment, what makes this surprising is that I'd expect `psa_driver_wrapper_sign_hash_get_num_ops()` to behave the same as `mbedtls_psa_sign_hash_get_num_ops()` - the naming certainly suggests that.

So, while I now understand why the `+=` needs to happen somewhere, I don't think it should be here but rather inside `psa_driver_wrapper_sign_hash_get_num_ops()`.

(Alternativaly, we could decide that `driver_xxx_get_num_ops()` has different semantics from `psa_xxx_get_num_ops()` but that seems like looking for trouble - if we went this way, I'd strongly advocate using different names.)

Reply:
At the moment, both `mbedtls_psa_sign_hash_get_num_ops()` and 'psa_driver_wrapper_sign_hash_get_num_ops()' return a delta of "num ops done in last operation", but I am guessing that from the naming you did not expect this to be the case (as previously discussed, neither did I when building this).

If the += happens in the driver wrapper, then 'psa_driver_wrapper_sign_hash_get_num_ops()' will return an absolute value, and `mbedtls_psa_sign_hash_get_num_ops()` will return a delta, which seems, as you say not an ideal solution, as then they definitely do not behave the same.

I *could* solve this by making `mbedtls_psa_sign_hash_get_num_ops()` return an absolute value, by adding in the context ops_done value. This would cause errors if you called it more than once per operation, obviously, however the current implementation breaks if you don't call it each and every operation.

One big question here is what the driver version would return (absolute or delta), but obviously we cannot know this, as there are no drivers.

What do you think?

Reply:
Fixed by moving the += to the driver level, and changed docs.